xonsh/lexer.py

```diff
@@ -1,13 +1,307 @@
 from __future__ import print_function, unicode_literals
 import re
 import sys
+import tokenize
+
+from io import BytesIO
+from keyword import kwlist
 
 from ply import lex
-from ply.lex import TOKEN
+from ply.lex import TOKEN, LexToken
+
+# mapping from tokenize to PLY
+# some keys are (type, name) tuples (for specific, e.g., keywords)
+# some keys are just a type, for things like strings/names
+# values are always a PLY token type
+token_map = {}
+
+# keywords
+for kw in kwlist:
+    token_map[(tokenize.NAME, kw)] = kw.upper()
+
+# operators
+op_map = {
+        # punctuation
+        ',': 'COMMA', '.': 'PERIOD', ';': 'SEMI', ':': 'COLON',
+        '...': 'ELLIPSIS',
+        # basic operators
+        '+': 'PLUS', '-': 'MINUS', '*': 'TIMES', '/': 'DIVIDE',
+        '//': 'DOUBLEDIV', '%': 'MOD', '**': 'POW', '|': 'PIPE',
+        '&': 'AMPERSAND', '~': 'TILDE', '^': 'XOR', '<<': 'LSHIFT',
+        '>>': 'RSHIFT', '<': 'LT', '<=': 'LE', '>': 'GT', '>=': 'GE',
+        '==': 'EQ', '!=': 'NE', '->': 'RARROW',
+        # assignment operators
+        '=': 'EQUALS', '+=': 'PLUSEQUAL', '-=': 'MINUSEQUAL',
+        '*=': 'TIMESEQUAL', '/=': 'DIVEQUAL', '%=': 'MODEQUAL',
+        '**=': 'POWEQUAL', '<<=': 'LSHIFTEQUAL', '>>=': 'RSHIFTEQUAL',
+        '&=': 'AMPERSANDEQUAL', '^=': 'XOREQUAL', '|=': 'PIPEEQUAL',
+        '//=': 'DOUBLEDIVEQUAL',
+}
+for (op, type) in op_map.items():
+    token_map[(tokenize.OP, op)] = type
+
+token_map[tokenize.NAME] = 'NAME'
+token_map[tokenize.NUMBER] = 'NUMBER'
+token_map[tokenize.STRING] = 'STRING'
+
+
+def handle_indent(state, token, stream):
+    level = len(token.string)
+    state['last'] = token
+    if token.type == tokenize.DEDENT:
+        state['indents'].pop()
+        yield _new_token('DEDENT', ' '*state['indents'][-1], token.start)
+    elif token.type == tokenize.INDENT:
+        state['indents'].append(level)
+        yield _new_token('INDENT', token.string, token.start)
+
```

Comment:
Shouldn't the rest of this function body be in an else-block to prevent the following code from being accessed accidentally after an INDENT or DEDENT.  In fact, if you did this, I think you might be able to get rid of the recursion and simply have:

```python
else:
    yield token
```